### ISSUE TITLE
Stop the madness, make something better.

### ISSUE THREAD

**Issue by ewtoombs** [radare/radare2#7178]
After a short, really fun, really frustrating time working with radare, I've had enough. This is a call to you to abandon radare2's current code base entirely, and the shoddy soft eng practices that birthed it, break compatibility with old scripts, and _make something better_. Let's start on a positive note.

## radare is still extremely awesome.
Yes, awesomeness is its sole redeemable feature. The idea of open source, arbitrarily modifiable, arbitrarily scriptable binary and assembly munging is amazing, and for all its many faults, radare has still shown its potential to me very clearly, and it has great potential. I want to see this, or something else like this succeed. OK, let's begin.

## The code is shit.
Functionality existing in actually-good third party libraries is duplicated in-house, poorly, for no reason, at extreme maintenance cost. Documentation of code is not only optional—it is not allowed. As a result, there are few to no clearly defined interfaces between functions and mortals like me have to read an entire function before knowing for sure what the valid inputs are, what the possible return values are, and what the side effects are. And since every function has some kind of mistake or other in it, I have no way of knowing if the mistake is intended behaviour, because the intended behaviour was—you guessed it—never documented. All this with a language known for being a bug factory. Speaking of language.

## The radare scripting language is shit.
No nestable evaluation, since evaluations can only be delimited with backticks. No clear types. The only thing that could reasonably be called a variable (a flag) may hold integers and nothing else. There is dirty statefulness every-goddamned-where. It's all very reminiscent of PHP or javascript, other languages cobbled together out of spare parts and garbage out of necessity. Break compatibility with legacy radare behaviour and make something better, seriously.

## The functionality is shit.
Nothing Works. What does work is finnicky and extremely buggy, with unexpected, probably unintended side effects. At least I think it's unintended, but I'll never know since there's no god damned documentation. There was clearly no premeditated design to any of it. It's pretty clear to me the founders kind of took an idea and some old software whose behaviour they liked and wanted to replicate and had at it. Every aspect of its functionality is quirky and surprising in some way.

All that said, I have had a lot of fun with this terrible but still-pretty-damned-awesome tool but I've reached the limits of my patience with radare's quirkiness, bugginess, and general shittiness. So I'm switching to either IDA, Hopper, or that new thing called somethingmonkey. Binarymonkey? Whatever. That's it for me. I've had it.

**Comment by radare** [comment]
👋

**Comment by radare** [comment]
This issue belongs to https://github.com/radare/r2hate can you reopen the issues there?

**Comment by joxeankoret** [comment]
#MakeRadareGreatAgain

**Comment by rlaemmert** [comment]
who cares?

**Comment by PaquitoRiviera** [comment]
> who cares?

At least me. Here, let me show you how much by being extremely verbose.

Although I don't agree that radare2 should be abandoned (or even rewritten in another language), 95% of the rest is an accurate description of my experience with radare2 and I suspect many also for many silent others (not so silent, actually). The poor dev practices and random priorities  are maddening to me as well. I also think most users are all too familiar with the pattern of sitting down to solve a problem **using** r2 and ending up chasing a problem **in** r2 instead.

One problem he neglected (though this issue is a typical example) is the way substantial criticism is handled. Usually by some unsightly mixture of mocking the person, rephrasing the problem into something unrelated but easily disposed of, hand waving the problem away and closing the issue as quickly as possible. Now, we also have a shiny new repo wholly dedicated to sweeping all these issues under the rug as "hate". Itself another example that something has going horribly wrong with the project.

The reason users get so frustrated and vent (beyond all of the excellent reasons above) is that, like the OP says, we clearly see radare2's enormous promise. Which just makes it all the more infuriating when you actually begin to use it and see how badly it fails to live up to that promise. I've seen, used and written some good software and some bad, I've never met a project that at the same time got me so excited in concept and so outraged in practice.

> Fine, so stop whining and start sending PRs to make it better.

Exactly the point, because this mechanical response blames a lack of resources for a situation that is the result of ongoing systemic failures in development and community practices. Blaming the problem on users not contributing is just another way to avoid the issue. New contributors can't fix bad processes, they can only urge the core developers to fix them, and when they do, they get labeled haters.

Until the core issues with the way r2 is developed are addressed no amount of PRs will get radare2 out of the swamp. New problems will always be created faster than they are solved. What makes Linux a success story, an unbeatable development process. R2 gets this wrong and this is how you end up with angry users and a handful of core developers who feel they're doing all the work and only get yelled at for their trouble, burn out, and gradually their contributions dwindle. And everytime @radare (the project, and the person) treats another person this way, it shows how committed it is to keeping things broken.

Radare2 will have to fix this, sooner or later. Until you do, I can virtually guarantee that you you'll keep receiving this message in one form or another from new people as they come in. Someday, you'll decide to stop calling them haters so you avoid dealing with the problems, and instead realize that we *want* radare2 to succeed, that we *want* it to be better, and that you need to start fixing these problem to turn "haters" into valuable contributors. It's up to you how long before that happens.
 
tl;dr: [Fortune](https://github.com/radare/radare2/blob/1ab8dc27ea174c787994d2fe6f4024fea7ec5353/doc/fortunes#L161). 3 years ago, what made that sentiment accurate, even inevitable, are still very much what cripples radare2 today. Isn't it time yet?

**Comment by Maijin** [comment]
@PaquitoRiviera Why so many comments about development, so far you did 0 contributions on the project as far as your profile indicates...

**Comment by PaquitoRiviera** [comment]
> @PaquitoRiviera Why so many comments about development

1. because so many problems with development
2. because like OP, I'm a frustrated user and I trace the cause of my frustration to bade development practices in the project. Which can only be fixed by discussion and persuasion, not PRs.

> so far you did 0 contributions on the project as far as your profile indicates...

Irrelevent. The issues I (and others) raise are serious process flaws that need to be openly discusses and addressed. Neither I nor anyone else has to "qualify" in order to point these out, if they're valid and if our experience elsewhere makes us believe that fixing them would make radare2 ultimately better. If these comments do make a difference eventually, they're a far more significant contribution to the project in the long run than fixing a few buffer overflows. Or adding some new ones.

Also, so far your profile indicates that you're diligently closing all such issues but fixing none of them,  so please don't throw stones.That fact and and this latest attempt to discredit my comments because of "my profile" just seem like attempts to avoid dealing with the deep problems this project has. 

You and all the other developers (present and future) have the most to gain from fixing them, btw.
